src/self-update/README.md

```diff
@@ -0,0 +1,89 @@
+  * This is to reduce external dependencies and so reduce risc for failures due missing tools or wrong version of tools.
+
+# Expected Pre-Conditions 
+
+The device and it's thereon installed thin-edge has to follow precondition below to allow successful self-update:
+
+1) The current (old) thin-edge on the device must be installed with usual thin-edge installation, or at leat as usual thin-edge installation would do. That is to assure new thin-edge version that is about to be installed has all external and internal files, folders and resources (e.g. config files) available.
+2) Deployment of thin-edge and mosquitto must not be changed manually after initial installation (e.g. moving/renaming installed files, moving/renaming configuration files used by thin-edge, renaming users/groups created during installation, ...).
+
+3) Package Dependencies need to be solved using SW Management before self-update. This is since self-update will disable auto-dep solving to reduce additional risk during self-update.
+
+*Hint: Before executing self-update on devices in the field it shall be tested on a kind of golden device that equals the device(s) in the field in system-environment, configuration and thin-edge version.*
+
+
+# Execution Flow and involved Components
+![Sequence Diagram Update SW-list](images/self-update.drawio.svg)
+
+
+# Design Principles
+
+### Details about the SM Plugin
+
+* A designated Software Management module type - namely "tedge" - will be used manage the self-update.
+
+* The existing SM plugin "apt" will be used and extended to manage that module type "tedge". 
+  * A device owner can implement an own plugin for module type "tedge", e.g. to use any other package manager than APT for self-update. 
+
+* A softlink to be created in `/etc/tedge/sm-plugins/` named `tedge` that points to `apt` plugin, will make the SM agent aware.
+
+* The plugin managing module type "tedge" must take all packages of an update request at once from SM Agent (i.E. Plugin API command "update-list" must be used). That is to get the plugin independent from potential restarts of the SM agent (those restarts are part of updating the SM agent).
+
+* Before any package update for module type "tedge" starts, all needed installation resources for the entire update request must be downloaded (i.E. all \*.deb packages for APT). That is to get the plugin uncoupled from any potential network issue during update procedure.
+  * For all packages with some URL given in the update request the SM Agent manages download before plugin start, and provides local file paths to the plugin.<br/>
+    TODO: What happens if one package download fails in SM agent? Does plugin maybe need to check if all files are locally available?
+  * For all packages without any URL given in the update request the Package Manager (i.E. APT) will manage the download as part of the update. Therefor the plugin must instruct the package manager to first download **all** packages. 
```

Comment:
It might be good to say how. I guess this is a combination of the `--download-only` and `--no-download` options of `apt-get`, but I would prefer to have to details here.

Reply:
I detailed that aspect as below:

- For module type „tedge“ the plugin must instruct APT to first download all packages that are part of the entire update request. Therefore implementation of plugin command update-list calls APT separately for download and for installation.
   - First all packages are downloaded by calling APT with command line option `--download-only`.
   - When all packages were successfully downloaded packages are installed by calling APT again with option `--no-download`.
   - If one or more downloads fail, the plugin will not install any package, but return with error.